test_tabular_write_callable_kb_interrupt_in_exit fails

initially ran into it while preparing a debian package for pyout, thought it might be some side-effect of pybuild environment. But seems to replicate on a fresh clone:

```
lena:~deb
$> git clone https://github.com/pyout/pyout pyout-upstream
Cloning into 'pyout-upstream'...
remote: Enumerating objects: 163, done.
remote: Counting objects: 100% (163/163), done.
remote: Compressing objects: 100% (61/61), done.
remote: Total 1821 (delta 114), reused 140 (delta 102), pack-reused 1658
Receiving objects: 100% (1821/1821), 401.38 KiB | 3.24 MiB/s, done.
Resolving deltas: 100% (1302/1302), done.
1 21615.....................................:Wed 29 Jan 2020 11:26:34 AM EST:.
lena:~deb
$> cd pyout-upstream
CHANGELOG.md  COPYING  README.rst  appveyor.yml  pyout/  pytest.ini  setup.cfg  setup.py  tox.ini
1 21616.....................................:Wed 29 Jan 2020 11:26:35 AM EST:.
(git)lena:~deb/pyout-upstream[master]
$> python3.7 -m pytest
======================================================================================================= test session starts =======================================================================================================
platform linux -- Python 3.7.5, pytest-4.6.6, py-1.8.0, pluggy-0.13.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/yoh/deb/gits/pyout-upstream/.hypothesis/examples')
rootdir: /home/yoh/deb/gits/pyout-upstream, inifile: pytest.ini
plugins: hypothesis-3.71.11, forked-1.0.2, timeout-1.3.3, xdist-1.30.0
collected 154 items                                                                                                                                                                                                               

pyout/tabular.py .                                                                                                                                                                                                          [  0%]
pyout/tests/test_elements.py .....                                                                                                                                                                                          [  3%]
pyout/tests/test_field.py .......                                                                                                                                                                                           [  8%]
pyout/tests/test_interface.py ....                                                                                                                                                                                          [ 11%]
pyout/tests/test_summary.py ....                                                                                                                                                                                            [ 13%]
pyout/tests/test_tabular.py ....................................................................................F........................                                                                                   [ 84%]
pyout/tests/test_tabular_dummy.py ....                                                                                                                                                                                      [ 87%]
pyout/tests/test_tests_utils.py ....                                                                                                                                                                                        [ 89%]
pyout/tests/test_truncate.py ................                                                                                                                                                                               [100%]

============================================================================================================ FAILURES =============================================================================================================
______________________________________________________________________________________ test_tabular_write_callable_kb_interrupt_during_wait _______________________________________________________________________________________

    @pytest.mark.timeout(10)
    def test_tabular_write_callable_kb_interrupt_during_wait():
        delay0 = Delayed("v0")
        delay1 = Delayed("v1")
    
        out = Tabular(max_workers=1)
    
        def run_tabular():
            def raise_kbint():
                raise KeyboardInterrupt
    
            out.wait = raise_kbint
            with out:
                out(OrderedDict([("name", "foo"), ("status", delay0.run)]))
                out(OrderedDict([("name", "bar"), ("status", delay1.run)]))
    
        thread = threading.Thread(target=run_tabular)
        thread.daemon = True
        thread.start()
        delay0.now = True
        delay1.now = True
        thread.join()
        stdout = out.stdout
        assert_contains_nc(stdout.splitlines(),
                           "foo v0",
>                          "bar   ")

pyout/tests/test_tabular.py:1271: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

collection = ['foo', 'bar', '', 'Keyboard interrupt registered', 'Canceled pending asynchronous workers. 1 worker already running'], items = ('foo v0', 'bar   '), kwargs = {}, count = 1

    def assert_contains(collection, *items, **kwargs):
        """Check that each item in `items` is in `collection`.
    
        Parameters
        ----------
        collection : list
        *items : list
            Query items.
        count : int, optional
            The number of times each item should occur in `collection`.
        cmp : callable, optional
            Function to compare equality with.  It will be called with the element
            from `items` as the first argument and the element from `collection` as
            the second.
    
        Raises
        ------
        AssertionError if any item does not occur `count` times in `collection`.
        """
        count = kwargs.pop("count", 1)
        cmp = kwargs.pop("cmp", eq)
        for item in items:
            if not len([x for x in collection if cmp(x, item)]) == count:
                raise AssertionError("{!r} (x{}) not in {!r}".format(
>                   item, count, collection))
E               AssertionError: 'foo v0' (x1) not in ['foo', 'bar', '', 'Keyboard interrupt registered', 'Canceled pending asynchronous workers. 1 worker already running']

pyout/tests/utils.py:28: AssertionError
------------------------------------------------------------------------------------------------------ Captured stderr call -------------------------------------------------------------------------------------------------------
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/yoh/deb/gits/pyout-upstream/pyout/tests/test_tabular.py", line 1260, in run_tabular
    out(OrderedDict([("name", "bar"), ("status", delay1.run)]))
  File "/home/yoh/deb/gits/pyout-upstream/pyout/interface.py", line 201, in __exit__
    failed = self.wait()
  File "/home/yoh/deb/gits/pyout-upstream/pyout/tests/test_tabular.py", line 1255, in raise_kbint
    raise KeyboardInterrupt
KeyboardInterrupt

============================================================================================== 1 failed, 153 passed in 8.79 seconds ===============================================================================================
python3.7 -m pytest  5.75s user 0.11s system 63% cpu 9.291 total
```
Here is the reproman trace which could be used to consult for versions of stuff used: http://www.onerussian.com/tmp/pyout-tests-trace-20200129.yml

<details>
<summary>and here is a full protocol of running it, although failing for a different reason</summary> 

```shell
(git)lena:~deb/pyout-upstream[master]
$> reproman execute --trace -- python3.7 -m pytest
Enter a resource name or ID: localshell

Some files were read and then written. We will only pack the final version of the file; reproducible experiments shouldn't change their input files
Configuration file written in /home/yoh/.cache/reproman/traces/20200129112850-ed/config.yml
Edit that file then run the packer -- use 'reprozip pack -h' for help
2020-01-29 11:28:54,300 [INFO   ] reading spec file /home/yoh/.cache/reproman/traces/20200129112850-ed/config.yml 
2020-01-29 11:28:54,518 [INFO   ] Entering iteration #1 over Tracers 
2020-01-29 11:28:56,439 [INFO   ] DebTracer: 1 envs with 149 other files remaining 
2020-01-29 11:28:56,441 [INFO   ] RPMTracer: 0 envs with 553 other files remaining 
2020-01-29 11:29:03,559 [INFO   ] CondaTracer: 0 envs with 553 other files remaining 
2020-01-29 11:29:07,852 [INFO   ] VenvTracer: 1 envs with 494 other files remaining 
2020-01-29 11:29:23,021 [INFO   ] VCSTracer: 1 envs with 456 other files remaining 
2020-01-29 11:29:25,551 [INFO   ] DockerTracer: 0 envs with 88 other files remaining 
2020-01-29 11:29:44,632 [INFO   ] SingularityTracer: 0 envs with 456 other files remaining 
2020-01-29 11:29:44,633 [INFO   ] Entering iteration #2 over Tracers 
2020-01-29 11:29:45,549 [INFO   ] DebTracer: 1 envs with 48 other files remaining 
2020-01-29 11:29:45,550 [INFO   ] RPMTracer: 0 envs with 416 other files remaining 
2020-01-29 11:29:50,634 [INFO   ] CondaTracer: 0 envs with 416 other files remaining 
2020-01-29 11:29:51,389 [INFO   ] VenvTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:04,583 [INFO   ] VCSTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:05,916 [INFO   ] DockerTracer: 0 envs with 48 other files remaining 
2020-01-29 11:30:25,437 [INFO   ] SingularityTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:25,437 [INFO   ] Entering iteration #3 over Tracers 
2020-01-29 11:30:25,797 [INFO   ] DebTracer: 0 envs with 48 other files remaining 
2020-01-29 11:30:25,800 [INFO   ] RPMTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:31,299 [INFO   ] CondaTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:32,103 [INFO   ] VenvTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:46,315 [INFO   ] VCSTracer: 0 envs with 416 other files remaining 
2020-01-29 11:30:47,794 [INFO   ] DockerTracer: 0 envs with 48 other files remaining 
2020-01-29 11:31:07,135 [INFO   ] SingularityTracer: 0 envs with 416 other files remaining 
2020-01-29 11:31:07,135 [INFO   ] No more changes or files to track.  Exiting the loop 
2020-01-29 11:31:07,515 [INFO   ] ReproMan trace /home/yoh/.cache/reproman/traces/20200129112850-ed/reproman.yml 
2020-01-29 11:31:07,515 [INFO   ] Executed the python3.7 command in the environment localshell 
============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-4.6.6, py-1.8.0, pluggy-0.13.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/yoh/deb/gits/pyout-upstream/.hypothesis/examples')
rootdir: /home/yoh/deb/gits/pyout-upstream, inifile: pytest.ini
plugins: hypothesis-3.71.11, forked-1.0.2, timeout-1.3.3, xdist-1.30.0
collected 154 items

pyout/tabular.py .                                                       [  0%]
pyout/tests/test_elements.py .....                                       [  3%]
pyout/tests/test_field.py .......                                        [  8%]
pyout/tests/test_interface.py ....                                       [ 11%]
pyout/tests/test_summary.py ....                                         [ 13%]
pyout/tests/test_tabular.py ............................................ [ 42%]
..........................2020-01-29 11:31:07,515 [ERROR  ] Command python3.7 failed to run in localshell: Failed to run ['/home/yoh/.cache/reproman/tracers/d8561c1bc528592b21c0e28d6f32c0a4/reproman_trace', '--logfile', '/home/yoh/.cache/reproman/traces/20200129112850-ed/tracer.log', '--dbfile', '/home/yoh/.cache/reproman/traces/20200129112850-ed/trace.sqlite3', '--', 'python3.7', '-m', 'pytest'] under None. Exit code=11. out============================== test session starts ==============================
platform linux -- Python 3.7.2, pytest-4.6.6, py-1.8.0, pluggy-0.13.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/yoh/deb/gits/pyout-upstream/.hypothesis/examples')
rootdir: /home/yoh/deb/gits/pyout-upstream, inifile: pytest.ini
plugins: hypothesis-3.71.11, forked-1.0.2, timeout-1.3.3, xdist-1.30.0
collected 154 items

pyout/tabular.py .                                                       [  0%]
pyout/tests/test_elements.py .....                                       [  3%]
pyout/tests/test_field.py .......                                        [  8%]
pyout/tests/test_interface.py ....                                       [ 11%]
pyout/tests/test_summary.py ....                                         [ 13%]
pyout/tests/test_tabular.py ............................................ [ 42%]
.......................... err= 
reproman execute --trace -- python3.7 -m pytest  105.75s user 47.24s system 107% cpu 2:21.74 total
```
</details>

and quick ideas on what could be a culprit?

FWIW, preliminary debian packaging for pyout  is pushed to https://salsa.debian.org/python-team/modules/pyout

failure shows itself on [OSX](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=140760&view=logs&j=c572e5d5-7ba7-55ce-5575-7ce1d9f7161e&t=81967b5f-0283-5c0b-87c1-e8fdeef842a3) and [Linux](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=140760&view=logs&j=10bda42a-d420-5c28-a43f-bc4e66405e5c&t=bef775d0-a1a2-55a3-9721-17147029baf0) while trying to build conda-forge package for pyout: https://github.com/conda-forge/staged-recipes/pull/11234 . Here is an example traceback of failure:
```
2020-04-07T03:02:53.2668260Z =================================== FAILURES ===================================
2020-04-07T03:02:53.2677490Z _______________ test_tabular_write_callable_kb_interrupt_in_exit _______________
2020-04-07T03:02:53.2680740Z 
2020-04-07T03:02:53.2681150Z     @pytest.mark.timeout(10)
2020-04-07T03:02:53.2681980Z     def test_tabular_write_callable_kb_interrupt_in_exit():
2020-04-07T03:02:53.2683140Z         delay0 = Delayed("v0")
2020-04-07T03:02:53.2684130Z         delay1 = Delayed("v1")
2020-04-07T03:02:53.2685180Z     
2020-04-07T03:02:53.2686360Z         out = Tabular(max_workers=1)
2020-04-07T03:02:53.2687010Z     
2020-04-07T03:02:53.2687490Z         def run_tabular():
2020-04-07T03:02:53.2688040Z             with out:
2020-04-07T03:02:53.2688720Z                 out(OrderedDict([("name", "foo"), ("status", delay0.run)]))
2020-04-07T03:02:53.2689600Z                 out(OrderedDict([("name", "bar"), ("status", delay1.run)]))
2020-04-07T03:02:53.2690410Z                 raise KeyboardInterrupt
2020-04-07T03:02:53.2691050Z     
2020-04-07T03:02:53.2691740Z         thread = threading.Thread(target=run_tabular)
2020-04-07T03:02:53.2692440Z         thread.daemon = True
2020-04-07T03:02:53.2693170Z         thread.start()
2020-04-07T03:02:53.2693840Z         delay0.now = True
2020-04-07T03:02:53.2694490Z         thread.join()
2020-04-07T03:02:53.2695130Z         stdout = out.stdout
2020-04-07T03:02:53.2696270Z         assert "KeyboardInterrupt" in stdout
2020-04-07T03:02:53.2697010Z         assert_contains_nc(stdout.splitlines(),
2020-04-07T03:02:53.2697710Z                            "foo v0",
2020-04-07T03:02:53.2698680Z >                          "bar   ")
2020-04-07T03:02:53.2699620Z 
2020-04-07T03:02:53.2701570Z ../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.7/site-packages/pyout/tests/test_tabular.py:1243: 
2020-04-07T03:02:53.2703060Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2020-04-07T03:02:53.2703610Z 
2020-04-07T03:02:53.2704870Z collection = ['foo', 'bar', '', 'KeyboardInterrupt() raised', 'Canceled pending asynchronous workers. 1 worker already running']
2020-04-07T03:02:53.2706140Z items = ('foo v0', 'bar   '), kwargs = {}, count = 1
2020-04-07T03:02:53.2706590Z 
2020-04-07T03:02:53.2707010Z     def assert_contains(collection, *items, **kwargs):
2020-04-07T03:02:53.2707600Z         """Check that each item in `items` is in `collection`.
2020-04-07T03:02:53.2708050Z     
2020-04-07T03:02:53.2708370Z         Parameters
2020-04-07T03:02:53.2709170Z         ----------
2020-04-07T03:02:53.2709600Z         collection : list
2020-04-07T03:02:53.2710030Z         *items : list
2020-04-07T03:02:53.2710430Z             Query items.
2020-04-07T03:02:53.2710900Z         count : int, optional
2020-04-07T03:02:53.2711460Z             The number of times each item should occur in `collection`.
2020-04-07T03:02:53.2712030Z         cmp : callable, optional
2020-04-07T03:02:53.2712660Z             Function to compare equality with.  It will be called with the element
2020-04-07T03:02:53.2713420Z             from `items` as the first argument and the element from `collection` as
2020-04-07T03:02:53.2713990Z             the second.
2020-04-07T03:02:53.2714320Z     
2020-04-07T03:02:53.2714650Z         Raises
2020-04-07T03:02:53.2715700Z         ------
2020-04-07T03:02:53.2716390Z         AssertionError if any item does not occur `count` times in `collection`.
2020-04-07T03:02:53.2717070Z         """
2020-04-07T03:02:53.2717620Z         count = kwargs.pop("count", 1)
2020-04-07T03:02:53.2718220Z         cmp = kwargs.pop("cmp", eq)
2020-04-07T03:02:53.2718780Z         for item in items:
2020-04-07T03:02:53.2719430Z             if not len([x for x in collection if cmp(x, item)]) == count:
2020-04-07T03:02:53.2720240Z                 raise AssertionError("{!r} (x{}) not in {!r}".format(
2020-04-07T03:02:53.2720970Z >                   item, count, collection))
2020-04-07T03:02:53.2722650Z E               AssertionError: 'foo v0' (x1) not in ['foo', 'bar', '', 'KeyboardInterrupt() raised', 'Canceled pending asynchronous workers. 1 worker already running']
2020-04-07T03:02:53.2723580Z 
2020-04-07T03:02:53.2725250Z ../_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.7/site-packages/pyout/tests/utils.py:28: AssertionError
2020-04-07T03:02:53.2726620Z =========================== short test summary info ============================
2020-04-07T03:02:53.2727520Z FAILED tests/test_tabular.py::test_tabular_write_callable_kb_interrupt_in_exit
2020-04-07T03:02:53.2728400Z ======================== 1 failed, 152 passed in 9.70s =========================
```
please investigate, since it blocks finishing packaging for debian and conda, and then of the using pyout packages.

--- a/pyout/tests/test_tabular.py
+++ b/pyout/tests/test_tabular.py
@@ -1229,6 +1229,10 @@ def test_tabular_write_callable_kb_interrupt_in_exit():
         with out:
             out(OrderedDict([("name", "foo"), ("status", delay0.run)]))
             out(OrderedDict([("name", "bar"), ("status", delay1.run)]))
+            # Hold up until output from the first callable has been
+            # written.
+            while "v0" not in out.stdout:
+                time.sleep(0.1)
             raise KeyboardInterrupt
 
     thread = threading.Thread(target=run_tabular)
@@ -1252,6 +1256,13 @@ def test_tabular_write_callable_kb_interrupt_during_wait():
 
     def run_tabular():
         def raise_kbint():
+            # Hold up until output from the callables has been
+            # written.
+            while True:
+                stdout = out.stdout
+                if "v0" and "v1" in stdout:
+                    break
+                time.sleep(0.1)
             raise KeyboardInterrupt
 
         out.wait = raise_kbint

